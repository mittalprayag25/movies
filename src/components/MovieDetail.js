import { withStyles } from '@material-ui/core/styles';
import { useState } from 'react';
import Grid from '@material-ui/core/Grid';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Typography from '@material-ui/core/Typography';
import Rating from '@material-ui/lab/Rating';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import CloseIcon from '@material-ui/icons/Close';
import moment from 'moment';
import classNames from 'classnames';
import CastCard from './CastCard';

const styles = {
  rootContainter: {
    marginTop: 20,
  },
  card: {
    maxWidth: '70%',
    margin: '0 auto',
  },
  topBilledCastsCard: {
    padding: 5,
  },
  media: {
    maxWidth: 300,
    objectFit: 'cover',
  },
  ratingContainer: {
    width: 300,
    height: 300,
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-evenly',
    alignItems: 'center',
  },
  closeIcon: {
    position: 'absolute',
    top: '10px',
    right: '10px',
  },
};

export default withStyles(styles)(
  ({
    classes,
    browse,
    movie: {
      title,
      overview,
      poster,
      released,
      runtime,
      updateRating,
      rating,
      credits: { casts = [] } = {},
    } = {},
  }) => {
    const [open, setOpen] = useState(rating === 0 || !rating);

    const changeRating = (num) => {
      setOpen(false);
      if (num) {
        updateRating(num);
      }
    };

    const close = () => {
      setOpen(false);
    };

    return (
      <Grid container spacing={6} className={classes.rootContainter}>
        <Grid item xs={12}>
          <Card elevation={0} className={classes.card}>
            <Button variant='contained' color='primary' onClick={browse}>
              Back To Movie List
            </Button>
          </Card>
        </Grid>
        <Grid item xs={12}>
          <Card className={classes.card}>
            <Grid container>
              <Grid item>
                <CardMedia
                  component='img'
                  alt={title}
                  className={classes.media}
                  image={poster}
                  title={title}
                />
              </Grid>
              <Grid item xs>
                <CardContent>
                  <Typography variant='h3'>{title}</Typography>
                  <Typography variant='overline' gutterBottom>
                    {moment(released).isValid()
                      ? moment(released).format('MMMM DD, YYYY')
                      : 'Unknown release date'}
                  </Typography>
                  <Typography variant='h6'>Runtime</Typography>
                  <Typography gutterBottom>
                    {Math.floor(runtime / 60)}h {runtime % 60}m
                  </Typography>
                  <Typography variant='h6'>Overview</Typography>
                  <Typography gutterBottom>{overview}</Typography>
                  {rating > 0 ? (
                    <Rating
                      name='simple-controlled'
                      value={rating}
                      onChange={(event, newValue) => changeRating(newValue)}
                    />
                  ) : null}
                </CardContent>
              </Grid>
            </Grid>
          </Card>
        </Grid>
        <Grid item xs={12}>
          <Card
            elevation={0}
            className={classNames(classes.card, classes.topBilledCastsCard)}
          >
            <Grid container spacing={5}>
              <Grid item xs={12}>
                <Typography variant='h5'>Top Billed Cast</Typography>
              </Grid>
              <Grid item xs={12} container spacing={5}>
                {casts.map((cast) => (
                  <Grid key={cast.id} item>
                    <CastCard cast={cast} />
                  </Grid>
                ))}
              </Grid>
            </Grid>
          </Card>
        </Grid>

        <Dialog open={open} data-testid='movie-detail-dialog'>
          <CardContent className={classes.ratingContainer}>
            <CloseIcon
              className={classes.closeIcon}
              onClick={close}
              data-testid='close-icon'
            />
            <Typography variant='h6'>Did you like the movie?</Typography>
            <Rating
              name='simple-controlled'
              data-testid='movie-detail-ratings'
              value={rating}
              onChange={(event, newValue) => changeRating(newValue)}
            />
          </CardContent>
        </Dialog>
      </Grid>
    );
  },
);
