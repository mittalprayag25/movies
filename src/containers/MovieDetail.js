import { useState, useEffect } from 'react';
import { withRouter } from 'react-router';
import * as R from 'ramda';
import {
  withResourcesGetters,
  useResources,
  resourceTypes,
  gettersOf,
} from 'with-resources';
import { withStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import CircularProgress from '@material-ui/core/CircularProgress';
import MovieDetail from '../components/MovieDetail';
import { withMovieRetrieveOneResource } from '../data/resources/movie';
import withMovieReview from '../data/resources/review';

const TOP_CASTS = 5;

const styles = {
  overlay: {
    position: 'fixed',
    top: 0,
    left: 0,
    background: 'rgba(0, 0, 0, 0.6)',
    width: '100vw',
    height: '100vh',
    zIndex: 999,
  },
  loading: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)',
  },
};

const padToLongestName = (casts) => {
  const longest = (name) => R.reduce(
    (max, item) => Math.max(max, R.pipe(R.prop(name), R.length)(item)),
    0,
  );
  const longestName = longest('name')(casts);
  const longestCharacter = longest('character')(casts);
  return casts.map(({ name, character, ...rest }) => ({
    ...rest,
    name: name.padEnd(longestName, ' '),
    character: character.padEnd(longestCharacter, ' '),
  }));
};

const MovieDetailContainer = ({
  classes,
  history,
  match: { params: { movieId } = {} },
}) => {
  const [ratingVal, setRatingVal] = useState(0);

  const updateRating = (rating) => {
    setRatingVal(rating);
    let ratedMovies = [];
    const savedMoviesWithRating = getSavedMoviesWithRating();
    if (savedMoviesWithRating !== null && savedMoviesWithRating.length >= 0) {
      ratedMovies = savedMoviesWithRating;
    }

    const movieObj = movie;
    /* Check if movie has been rated earlier */
    if (!Object.prototype.hasOwnProperty.call(movieObj, 'rating')) {
      ratedMovies.push({ movieId, rating });
      saveMoviesWithRating(ratedMovies);
      return;
    }
    const index = R.findIndex(R.propEq('movieId', movie.id.toString()))(
      ratedMovies,
    );
    if (index > -1) {
      updateRatings(ratedMovies, rating, index);
    }
  };

  const updateRatings = (ratedMovies, rating, index) => {
    const movieAtIndex = ratedMovies[index];
    movieAtIndex.rating = rating;
    ratedMovies.splice(index, 1);
    ratedMovies.push(movieAtIndex);
    saveMoviesWithRating(ratedMovies);
  };

  const saveMoviesWithRating = (movies) => {
    localStorage.setItem('ratedMovies', JSON.stringify(movies));
  };

  const getSavedMoviesWithRating = () => JSON.parse(localStorage.getItem('ratedMovies'));

  const getModifiedMovieObject = (rawMovie) => {
    const savedMoviesWithRating = getSavedMoviesWithRating();
    if (!savedMoviesWithRating) {
      return rawMovie;
    }

    const rawMovieId = rawMovie.id.toString();
    const find = (obj) => obj.movieId === rawMovieId;
    const filterMovie = (data) => {
      const obj = { ...rawMovie, rating: data.rating };
      return obj;
    };

    const obj = R.map(filterMovie, R.filter(find, savedMoviesWithRating))[0]; // should not have more then one value though
    return obj || rawMovie;
  };

  const movieResource = useResources([withMovieRetrieveOneResource(movieId)]);

  const { actionCreators } = useResources([
    withMovieReview(ratingVal, movieId),
  ]);

  useEffect(() => {
    if (ratingVal > 0) {
      actionCreators.review.ajax({
        cargo: {
          resourceType: resourceTypes.REVIEW,
          method: 'retrieveOne',
          input: {
            params: {
              ids: [{ name: 'movieId', value: movieId }],
            },
            queries: [
              {
                name: 'guest_session_id',
                value: localStorage.getItem('guest_session_id'),
              },
            ],
            content: { value: ratingVal },
          },
        },
      });
    }
  }, [ratingVal]);

  let { movie } = gettersOf(resourceTypes.MOVIES).getMovie()(
    movieResource.data,
  );

  if (movie) {
    // If defined then modify the movie with rating
    movie = { ...movie, updateRating: (rating) => updateRating(rating) };
    movie = getModifiedMovieObject(movie); // Modified Movie Object with rating added
  }

  movie = R.pipe(
    R.juxt([
      R.pipe(
        R.pathOr([], ['credits', 'casts']),
        R.take(TOP_CASTS),
        padToLongestName,
        R.map(({ id, ...rest }) => ({
          ...rest,
          id,
          view: () => history.push(`/cast/${id}`),
        })),
      ),
      R.identity,
    ]),
    R.apply(R.assocPath(['credits', 'casts'])),
  )(movie);

  return R.prop(
    'loading',
    gettersOf(resourceTypes.MOVIES).getStatus()('retrieveOne')(
      movieResource.data,
    ),
  ) ? (
    <Paper className={classes.overlay}>
      <CircularProgress className={classes.loading} />
    </Paper>
    ) : (
    <MovieDetail movie={movie} browse={() => history.push('/')} />
    );
};

export default R.compose(
  withRouter,
  withResourcesGetters,
  withStyles(styles),
)(MovieDetailContainer);
