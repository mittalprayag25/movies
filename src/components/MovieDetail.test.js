import { render, fireEvent, screen } from '@testing-library/react';
import MovieDetail from './MovieDetail';
import '@testing-library/jest-dom/extend-expect';

describe('The MovieDetail component', async () => {
  it('should render movie title and release date', () => {
    const { container } = render(
      <MovieDetail
        movie={{
          title: 'I am a test movie',
          poster:
            'https://image.tmdb.org/t/p/w600_and_h900_bestv2/pgqgaUx1cJb5oZQQ5v0tNARCeBp.jpg',
          released: new Date('March 24, 2021'),
          runtime: 120,
          rating: 0,
        }}
      />,
      {},
    );
    expect(container).toMatchSnapshot();
    fireEvent.click(screen.getByTestId('close-icon'));
  });

  it('should render Unknown release date with incorrect date', () => {
    const { container } = render(
      <MovieDetail
        movie={{
          title: 'I am a test movie',
          poster:
            'https://image.tmdb.org/t/p/w600_and_h900_bestv2/pgqgaUx1cJb5oZQQ5v0tNARCeBp.jpg',
          released: 'asdf',
          rating: 2,
        }}
      />,
      {},
    );
    expect(container).toMatchSnapshot();
  });
});
