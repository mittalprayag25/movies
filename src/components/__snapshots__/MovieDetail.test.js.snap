// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The MovieDetail component should render Unknown release date with incorrect date 1`] = `
<div>
  <div
    class="MuiGrid-root Component-rootContainter-7 MuiGrid-container MuiGrid-spacing-xs-6"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiPaper-root MuiCard-root Component-card-8 MuiPaper-elevation0 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Back To Movie List
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiPaper-root MuiCard-root Component-card-8 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <img
              alt="I am a test movie"
              class="MuiCardMedia-root Component-media-10 MuiCardMedia-media MuiCardMedia-img"
              src="https://image.tmdb.org/t/p/w600_and_h900_bestv2/pgqgaUx1cJb5oZQQ5v0tNARCeBp.jpg"
              title="I am a test movie"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
          >
            <div
              class="MuiCardContent-root"
            >
              <h3
                class="MuiTypography-root MuiTypography-h3"
              >
                I am a test movie
              </h3>
              <span
                class="MuiTypography-root MuiTypography-overline MuiTypography-gutterBottom"
              >
                Unknown release date
              </span>
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Runtime
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
              >
                NaN
                h 
                NaN
                m
              </p>
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Overview
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
              />
              <span
                class="MuiRating-root"
              >
                <label
                  class="MuiRating-label"
                  for="simple-controlled-1"
                >
                  <span
                    class="MuiRating-icon MuiRating-iconFilled"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiRating-visuallyhidden"
                  >
                    1 Star
                  </span>
                </label>
                <input
                  class="MuiRating-visuallyhidden"
                  id="simple-controlled-1"
                  name="simple-controlled"
                  type="radio"
                  value="1"
                />
                <label
                  class="MuiRating-label"
                  for="simple-controlled-2"
                >
                  <span
                    class="MuiRating-icon MuiRating-iconFilled"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiRating-visuallyhidden"
                  >
                    2 Stars
                  </span>
                </label>
                <input
                  checked=""
                  class="MuiRating-visuallyhidden"
                  id="simple-controlled-2"
                  name="simple-controlled"
                  type="radio"
                  value="2"
                />
                <label
                  class="MuiRating-label"
                  for="simple-controlled-3"
                >
                  <span
                    class="MuiRating-icon MuiRating-iconEmpty"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiRating-visuallyhidden"
                  >
                    3 Stars
                  </span>
                </label>
                <input
                  class="MuiRating-visuallyhidden"
                  id="simple-controlled-3"
                  name="simple-controlled"
                  type="radio"
                  value="3"
                />
                <label
                  class="MuiRating-label"
                  for="simple-controlled-4"
                >
                  <span
                    class="MuiRating-icon MuiRating-iconEmpty"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiRating-visuallyhidden"
                  >
                    4 Stars
                  </span>
                </label>
                <input
                  class="MuiRating-visuallyhidden"
                  id="simple-controlled-4"
                  name="simple-controlled"
                  type="radio"
                  value="4"
                />
                <label
                  class="MuiRating-label"
                  for="simple-controlled-5"
                >
                  <span
                    class="MuiRating-icon MuiRating-iconEmpty"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiRating-visuallyhidden"
                  >
                    5 Stars
                  </span>
                </label>
                <input
                  class="MuiRating-visuallyhidden"
                  id="simple-controlled-5"
                  name="simple-controlled"
                  type="radio"
                  value="5"
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiPaper-root MuiCard-root Component-card-8 Component-topBilledCastsCard-9 MuiPaper-elevation0 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5"
            >
              Top Billed Cast
            </h5>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5 MuiGrid-item MuiGrid-grid-xs-12"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`The MovieDetail component should render movie title and release date 1`] = `
<div
  aria-hidden="true"
>
  <div
    class="MuiGrid-root Component-rootContainter-1 MuiGrid-container MuiGrid-spacing-xs-6"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiPaper-root MuiCard-root Component-card-2 MuiPaper-elevation0 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Back To Movie List
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiPaper-root MuiCard-root Component-card-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <img
              alt="I am a test movie"
              class="MuiCardMedia-root Component-media-4 MuiCardMedia-media MuiCardMedia-img"
              src="https://image.tmdb.org/t/p/w600_and_h900_bestv2/pgqgaUx1cJb5oZQQ5v0tNARCeBp.jpg"
              title="I am a test movie"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
          >
            <div
              class="MuiCardContent-root"
            >
              <h3
                class="MuiTypography-root MuiTypography-h3"
              >
                I am a test movie
              </h3>
              <span
                class="MuiTypography-root MuiTypography-overline MuiTypography-gutterBottom"
              >
                March 24, 2021
              </span>
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Runtime
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
              >
                2
                h 
                0
                m
              </p>
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Overview
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiPaper-root MuiCard-root Component-card-2 Component-topBilledCastsCard-3 MuiPaper-elevation0 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5"
            >
              Top Billed Cast
            </h5>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5 MuiGrid-item MuiGrid-grid-xs-12"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
